### PR TITLE
add set_suppress_warnings() to libshaderc

### PR DESCRIPTION
**Pull request by Qining**
Maybe we should update all the 'Set' in shaderc.h to 'Sets'?

### CODE REVIEW

File: libshaderc/include/shaderc.h [hunk 1]
```diff
@@ -128,6 +128,10 @@ void shaderc_compile_options_set_disassembly_mode(
 void shaderc_compile_options_set_preprocessing_only_mode(
     shaderc_compile_options_t options);
 
+// Sets the compiler to suppress all warning messages before they are output.
```
Comment (dneto0) [inline libshaderc/include/shaderc.h]: Please change to "Sets the compiler mode to suppress warnings."

There are two fixes here:
1. add "mode" to make the grammar work.
2. recognize that there are two conceptual phases: generating a warning, and outputting it.  There's a distinction in the sense that someone could have "-Werror" which changes a *generated* warning into an error, but still not print out the warning as a warning.

Comment (dneto0) [inline libshaderc/include/shaderc.h]: In another CL we can fix the other "Sets the compiler to..." comments to be "Sets the compiler mode to..."

File: libshaderc/src/shaderc_cpp_test.cc [hunk 2]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
               HasSubstr("void main(){ }"));
 }
 
+TEST(CppInterface, WarningsOnLine) {
+  shaderc::Compiler compiler;
+  shaderc::CompileOptions options;
+  // By default the compiler will emit a warning on line 2 complaining
```
Comment (dneto0) [inline libshaderc/src/shaderc_cpp_test.cc]: Good comment!

File: libshaderc/src/shaderc_cpp_test.cc [hunk 3]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+  const std::string kDeprecatedAttributeShader =
+      "#version 130\n"
+      "attribute float x;\n"
+      "void main() {}\n";
+  shaderc::SpvModule result_suppress_warnings_on_line =
+      compiler.CompileGlslToSpv(kDeprecatedAttributeShader,
+                                shaderc_glsl_vertex_shader, options);
+  EXPECT_TRUE(result_suppress_warnings_on_line.GetSuccess());
+  EXPECT_THAT(
+      result_suppress_warnings_on_line.GetErrorMessage().c_str(),
```
Comment (dneto0) [inline libshaderc/src/shaderc_cpp_test.cc]: You probably don't need the c_str() here.

File: libshaderc/src/shaderc_cpp_test.cc [hunk 4]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
               HasSubstr("void main(){ }"));
 }
 
+TEST(CppInterface, WarningsOnLine) {
+  shaderc::Compiler compiler;
+  shaderc::CompileOptions options;
+  // By default the compiler will emit a warning on line 2 complaining
+  // 'float' as a deprecated attribute in version 130.
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: "as" -> "is"?

File: libshaderc/src/shaderc_cpp_test.cc [hunk 5]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+      "#version 550\n"
+      "void main() {}\n";
+  shaderc::SpvModule result_suppress_warnings_on_line =
+      compiler.CompileGlslToSpv(kMinimalUnknownVersionShader,
+                                shaderc_glsl_vertex_shader,
+                                options_suppress_warnings);
+  EXPECT_TRUE(result_suppress_warnings_on_line.GetSuccess());
+  EXPECT_STREQ(result_suppress_warnings_on_line.GetErrorMessage().c_str(), "");
+}
+
```
Comment (dneto0) [inline libshaderc/src/shaderc_cpp_test.cc]: Please add a test to check this option survives cloning the options object.

**Comment by dneto0** [comment]
-1.  Please fix the blurb wording.  Possibly minor cleanup with removal of c_str(), and add a test for cloning.

File: libshaderc/include/shaderc.h (hunk 1 above)
Comment (dekimir) [inline libshaderc/include/shaderc.h]: @dneto0 another CL = #60 .

File: libshaderc/src/shaderc_cpp_test.cc [hunk 6]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+TEST(CppInterface, WarningsOnLine) {
+  shaderc::Compiler compiler;
+  shaderc::CompileOptions options;
+  // By default the compiler will emit a warning on line 2 complaining
+  // 'float' as a deprecated attribute in version 130.
+  const std::string kDeprecatedAttributeShader =
+      "#version 130\n"
+      "attribute float x;\n"
+      "void main() {}\n";
+  shaderc::SpvModule result_suppress_warnings_on_line =
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: Make it `const`.

Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: Also, why "suppress" in the name?  I don't understand what's being suppressed here.

File: libshaderc/src/shaderc_cpp_test.cc [hunk 7]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+                "removed in future release\n"));
+}
+
+TEST(CppInterface, SuppressWarningsOnLine) {
+  shaderc::Compiler compiler;
+  shaderc::CompileOptions options_suppress_warnings;
+  // Sets the compiler to suppress warnings, so that the deprecated attribute
+  // warning won't be emitted.
+  options_suppress_warnings.SetSuppressWarnings();
+  const std::string kDeprecatedAttributeShader =
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: DRY

File: libshaderc/src/shaderc_cpp_test.cc [hunk 8]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+      "#version 130\n"
+      "attribute float x;\n"
+      "void main() {}\n";
+  shaderc::SpvModule result_suppress_warnings_on_line =
+      compiler.CompileGlslToSpv(kDeprecatedAttributeShader,
+                                shaderc_glsl_vertex_shader,
+                                options_suppress_warnings);
+  EXPECT_TRUE(result_suppress_warnings_on_line.GetSuccess());
+  EXPECT_STREQ(result_suppress_warnings_on_line.GetErrorMessage().c_str(), "");
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: Don't need `c_str()` or `STREQ`.

Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: Also, expected `""` should be the first argument.

File: libshaderc/src/shaderc_cpp_test.cc [hunk 9]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+  // By default the compiler will emit a warning as version 550 is an unknown
+  // version.
+  const std::string kMinimalUnknownVersionShader =
+      "#version 550\n"
+      "void main() {}\n";
+  shaderc::SpvModule result_suppress_warnings_on_line =
+      compiler.CompileGlslToSpv(kMinimalUnknownVersionShader,
+                                shaderc_glsl_vertex_shader, options);
+  EXPECT_TRUE(result_suppress_warnings_on_line.GetSuccess());
+  EXPECT_THAT(result_suppress_warnings_on_line.GetErrorMessage().c_str(),
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: `c_str()` unneeded

File: libshaderc/src/shaderc_cpp_test.cc [hunk 10]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+
+TEST(CppInterface, GlobalWarnings) {
+  shaderc::Compiler compiler;
+  shaderc::CompileOptions options;
+  // By default the compiler will emit a warning as version 550 is an unknown
+  // version.
+  const std::string kMinimalUnknownVersionShader =
+      "#version 550\n"
+      "void main() {}\n";
+  shaderc::SpvModule result_suppress_warnings_on_line =
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: Again, I don't understand why this variable has `suppress` in its name.

File: libshaderc/src/shaderc_cpp_test.cc [hunk 11]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+  shaderc::Compiler compiler;
+  shaderc::CompileOptions options_suppress_warnings;
+  // Sets the compiler to suppress warnings, so that the deprecated attribute
+  // warning won't be emitted.
+  options_suppress_warnings.SetSuppressWarnings();
+  const std::string kDeprecatedAttributeShader =
+      "#version 130\n"
+      "attribute float x;\n"
+      "void main() {}\n";
+  shaderc::SpvModule result_suppress_warnings_on_line =
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: Make it `const`.

File: libshaderc/src/shaderc_cpp_test.cc (hunk 10 above)
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: Make it `const`.

File: libshaderc/src/shaderc_cpp_test.cc [hunk 12]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+              HasSubstr("warning: version 550 is unknown.\n"));
+}
+
+TEST(CppInterface, SuppressGlobalWarnings) {
+  shaderc::Compiler compiler;
+  shaderc::CompileOptions options_suppress_warnings;
+  // Sets the compiler to suppress warnings, so that the unknown version warning
+  // won't be emitted.
+  options_suppress_warnings.SetSuppressWarnings();
+  const std::string kMinimalUnknownVersionShader =
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: DRY

File: libshaderc/src/shaderc_cpp_test.cc [hunk 13]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+      "#version 550\n"
+      "void main() {}\n";
+  shaderc::SpvModule result_suppress_warnings_on_line =
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: Make it `const`.

File: libshaderc/src/shaderc_cpp_test.cc [hunk 14]
```diff
@@ -326,6 +326,76 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+      compiler.CompileGlslToSpv(kMinimalUnknownVersionShader,
+                                shaderc_glsl_vertex_shader,
+                                options_suppress_warnings);
+  EXPECT_TRUE(result_suppress_warnings_on_line.GetSuccess());
+  EXPECT_STREQ(result_suppress_warnings_on_line.GetErrorMessage().c_str(), "");
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: Don't need `c_str()` or `STREQ`.  Expected value is the first argument, not second.

File: libshaderc/src/shaderc_cpp_test.cc (hunk 7 above)
Comment (dneto0) [inline libshaderc/src/shaderc_cpp_test.cc]: Qining:  DRY = Don't Repeat Yourself

File: libshaderc/src/shaderc_cpp_test.cc (hunk 8 above)
Comment (dneto0) [inline libshaderc/src/shaderc_cpp_test.cc]: Right.  These are confusing.  In EXPECT_<predicate> macros, the expected thing is the first argument, and the computed thing is the second argument.
In EXPECT_THAT the computed thing is the first arg, and the predicate is spelled out in the second argument.

**Comment by dneto0** [comment]
+1.  Would like @dekimir to weigh in.

File: libshaderc/include/shaderc.h [hunk 15]
```diff
@@ -113,7 +113,7 @@ void shaderc_compile_options_release(shaderc_compile_options_t options);
 void shaderc_compile_options_add_macro_definition(
     shaderc_compile_options_t options, const char* name, const char* value);
 
-// Set the compiler to emit a disassembly text instead of a binary. In
```
Comment (dekimir) [inline libshaderc/include/shaderc.h]: I'm not sure why this shows up as a diff in your PR.  I'd have expected it to be the same on both sides.

File: libshaderc/src/shaderc_cpp_test.cc [hunk 16]
```diff
@@ -30,6 +30,13 @@ const char kMinimalShader[] = "void main(){}";
 const std::string kMinimalShaderWithMacro =
     "#define E main\n"
     "void E(){}\n";
+const std::string kDeprecatedAttributeShader =
+    "#version 130\n"
+    "attribute float x;\n"
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: Put the comment from line 339 here, so the reader will always see it, no matter which method uses this.

File: libshaderc/src/shaderc_cpp_test.cc [hunk 17]
```diff
@@ -30,6 +30,13 @@ const char kMinimalShader[] = "void main(){}";
...
+    "void main() {}\n";
+const std::string kMinimalUnknownVersionShader =
+    "#version 550\n"
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: Put the comment from line 374 here.

File: libshaderc/src/shaderc_cpp_test.cc [hunk 18]
```diff
@@ -326,6 +333,74 @@ TEST(CppInterface, PreprocessingOnlyModeSecondOverridesDisassemblyMode) {
...
+  // warning won't be emitted.
+  options_suppress_warnings.SetSuppressWarnings();
+  const shaderc::SpvModule result_suppress_warnings_on_line =
+      compiler.CompileGlslToSpv(kDeprecatedAttributeShader,
+                                shaderc_glsl_vertex_shader,
+                                options_suppress_warnings);
+  EXPECT_TRUE(result_suppress_warnings_on_line.GetSuccess());
+  EXPECT_EQ("", result_suppress_warnings_on_line.GetErrorMessage());
+
+  shaderc::CompileOptions cloned_options(options_suppress_warnings);
```
Comment (dekimir) [inline libshaderc/src/shaderc_cpp_test.cc]: I'd make this a separate test case.

File: libshaderc/include/shaderc.h (hunk 15 above)
Comment (Qining) [inline libshaderc/include/shaderc.h]: I rebased my local branch to google's master branch and the master branch does the changes. But the remote branch stay unchanged, so this diff also shows some changes in master.

File: libshaderc/src/shaderc_test.cc [hunk 19]
```diff
@@ -72,7 +72,7 @@ class Compilation {
 
   ~Compilation() { shaderc_module_release(compiled_result_); }
 
-  shaderc_spv_module_t result() { return compiled_result_; }
```
Comment (dekimir) [inline libshaderc/src/shaderc_test.cc]: Lines 75-274 shouldn't be different.

File: libshaderc/src/shaderc_test.cc [hunk 20]
```diff
@@ -264,14 +264,78 @@ TEST(CompileStringWithOptions, PreprocessingOnlyOption) {
...
+      HasSubstr(":2: warning: attribute deprecated in version 130; may be "
+                "removed in future release\n"));
+}
+
+TEST(CompileStringWithOptions, SuppressWarningsOnLine) {
+  Compiler compiler;
+  compile_options_ptr options(shaderc_compile_options_initialize());
+  shaderc_compile_options_set_suppress_warnings(options.get());
+  ASSERT_NE(nullptr, compiler.get_compiler_handle());
+  const std::string kDeprecatedAttributeShader =
```
Comment (dekimir) [inline libshaderc/src/shaderc_test.cc]: DRY

File: libshaderc/src/shaderc_test.cc [hunk 21]
```diff
@@ -264,14 +264,78 @@ TEST(CompileStringWithOptions, PreprocessingOnlyOption) {
...
+  EXPECT_THAT(shaderc_module_get_error_message(comp.result()),
+              HasSubstr("version 550 is unknown.\n"));
+}
+
+TEST(CompileStringWithOptions, SuppressGlobalWarnings) {
+  Compiler compiler;
+  compile_options_ptr options(shaderc_compile_options_initialize());
+  shaderc_compile_options_set_suppress_warnings(options.get());
+  ASSERT_NE(nullptr, compiler.get_compiler_handle());
+  const std::string kMinimalUnknownVersionShader =
```
Comment (dekimir) [inline libshaderc/src/shaderc_test.cc]: DRY

File: libshaderc/src/shaderc_test.cc [hunk 22]
```diff
@@ -296,25 +360,24 @@ TEST(CompileStringWithOptions, TargetEnv) {
   // Confirm that this shader compiles with shaderc_target_env_opengl_compat;
   // if targeting Vulkan, glslang will fail to compile it
   const std::string kGlslShader =
-    R"(#version 100
```
Comment (dekimir) [inline libshaderc/src/shaderc_test.cc]: There should be no differences from this line on.

**Comment by Qining** [comment]
For some reason, this PR doesn't reflect google:master properly.  Closing and recreating.

**Pull request closed without merging**